exporter/opensearchexporter/model.go

```diff
@@ -0,0 +1,100 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package opensearchexporter // import "github.com/open-telemetry/opentelemetry-collector-contrib/exporter/opensearchexporter"
+
+import (
+	"bytes"
+	"encoding/json"
+
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/plog"
+	"go.opentelemetry.io/collector/pdata/ptrace"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/exporter/opensearchexporter/internal/objmodel"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/internal/coreinternal/traceutil"
+)
+
+type mappingModel interface {
+	encodeLog(pcommon.Resource, plog.LogRecord) ([]byte, error)
+	encodeSpan(pcommon.Resource, ptrace.Span) ([]byte, error)
+}
+
+// encodeModel tries to keep the event as close to the original open telemetry semantics as is.
+// No fields will be mapped by default.
+//
+// Field deduplication and dedotting of attributes is supported by the encodeModel.
+//
+// See: https://github.com/open-telemetry/oteps/blob/master/text/logs/0097-log-data-model.md
+type encodeModel struct {
+	dedup bool
+	dedot bool
+}
+
+const (
+	traceIDField   = "traceID"
+	spanIDField    = "spanID"
+	attributeField = "attribute"
+)
+
+func (m *encodeModel) encodeLog(resource pcommon.Resource, record plog.LogRecord) ([]byte, error) {
+	var document objmodel.Document
+	document.AddTimestamp("@timestamp", record.Timestamp()) // We use @timestamp in order to ensure that we can index if the default data stream logs template is used.
+	document.AddTraceID("TraceId", record.TraceID())
+	document.AddSpanID("SpanId", record.SpanID())
+	document.AddInt("TraceFlags", int64(record.Flags()))
+	document.AddString("SeverityText", record.SeverityText())
+	document.AddInt("SeverityNumber", int64(record.SeverityNumber()))
+	document.AddAttribute("Body", record.Body())
+	document.AddAttributes("Attributes", record.Attributes())
+	document.AddAttributes("Resource", resource.Attributes())
+
+	if m.dedup {
+		document.Dedup()
+	} else if m.dedot {
+		document.Sort()
+	}
+
+	var buf bytes.Buffer
+	err := document.Serialize(&buf, m.dedot)
+	return buf.Bytes(), err
+}
+
+func (m *encodeModel) encodeSpan(resource pcommon.Resource, span ptrace.Span) ([]byte, error) {
+	var document objmodel.Document
+	document.AddTimestamp("@timestamp", span.StartTimestamp()) // We use @timestamp in order to ensure that we can index if the default data stream logs template is used.
+	document.AddTimestamp("EndTimestamp", span.EndTimestamp())
+	document.AddTraceID("TraceId", span.TraceID())
+	document.AddSpanID("SpanId", span.SpanID())
+	document.AddSpanID("ParentSpanId", span.ParentSpanID())
+	document.AddString("Name", span.Name())
+	document.AddString("Kind", traceutil.SpanKindStr(span.Kind()))
+	document.AddInt("TraceStatus", int64(span.Status().Code()))
+	document.AddString("Link", spanLinksToString(span.Links()))
+	document.AddAttributes("Attributes", span.Attributes())
+	document.AddAttributes("Resource", resource.Attributes())
```

Comment:
This looks very different to the structure DataPrepper would generate. This is important, since the OpenSearch Observability Trace Analytics build on that data model. Here is an example from an OpenSearch index:

```json
"_source": {
    "kind": "SPAN_KIND_INTERNAL",
    "traceId": "eaa69f00b4ab63fd618eadffb00e60d7",
    "spanId": "75d7d735cb2b4972",
    "parentSpanId": "84f4ad4086932693",
    "traceState": "",
    "traceGroup": null,
    "serviceName": "sample",
    "name": "LogController.generateLog",
    "startTime": "2023-06-09T14:32:34.444486691Z",
    "endTime": "2023-06-09T14:32:34.446411315Z",
    "durationInNanos": 1924624,
    "status.code": 0,
    "events": [],
    "droppedEventsCount": 0,
    "traceGroupFields": {
      "endTime": null,
      "durationInNanos": null,
      "statusCode": null
    },
    "links": [],
    "droppedLinksCount": 0,
    "instrumentationScope.version": "1.25.0-alpha",
    "instrumentationScope.name": "io.opentelemetry.spring-webmvc-3.1",
    "resource.attributes.process@pid": 7,
    "resource.attributes.host@arch": "amd64",
    "resource.attributes.telemetry@sdk@version": "1.25.0",
    "resource.attributes.service@name": "sample",
    "resource.attributes.process@runtime@name": "Java(TM) SE Runtime Environment",
    "resource.attributes.os@type": "linux",
    "resource.attributes.telemetry@sdk@language": "java",
    "resource.attributes.host@name": "18232480-822d-40e2-5ded-ebb2",
    "resource.attributes.process@runtime@version": "8.1.094",
    "resource.attributes.telemetry@sdk@name": "opentelemetry",
    "resource.attributes.telemetry@auto@version": "1.25.0",
    "resource.attributes.os@description": "Linux 5.19.0-41-generic",
    "span.attributes.thread@id": 36,
    "span.attributes.thread@name": "http-nio-8080-exec-8",
    "droppedAttributesCount": 0
  }
```

How does the two naming schemes align in OpenSearch. Is there already support for both?

Reply:
@KarstenSchnitter yes this was based on the elastic exporter. I will be changing the schema shortly.

Reply:
@KarstenSchnitter could you please open a DataPrepper issue to make any necessary changes?

Reply:
@kkondaka what would be the concrete changes required in DataPrepper? It is not clear to me, what naming would be preferable. Some are just capitalization, some are specific to DataPrepper (extra serviceName and traceGroup) and others dilute understanding of the original OTel fields like `@timestamp` and `TraceStatus` (to a lesser extend). `Attributes` is in plural, where `Link` and `Resource` are not. 

I would prefer to not change the names in DataPrepper, since we use it in production. If this is necessary for alignment between the two tools, so be it.